lucet-runtime/lucet-runtime-internals/src/future.rs

```diff
@@ -0,0 +1,135 @@
+    /// used to run the VM, otherwise it will terminate the instance.
+    pub fn run_await<'a, R>(&'a self, f: impl Future<Output = R> + 'a) -> R
+    where
+        R: Any + 'static,
+    {
+        // Die if we aren't in Instance::run_async
+        match self.instance().state {
+            State::Running { async_context } => {
+                if !async_context {
+                    panic!(TerminationDetails::AwaitNeedsAsync)
+                }
+            }
+            _ => unreachable!("Access to vmctx implies instance is Running"),
+        }
+        // Wrap the Output of `f` as a boxed ResumeVal. Then, box the entire
+        // async computation.
+        let f = Box::pin(async move { ResumeVal(Box::new(f.await)) });
+        // Change the lifetime of the async computation from `'a` to `'static.
+        // We need to lie about this lifetime so that `YieldedFuture` may impl
+        // `Any` and be passed through the yield. `Instance::run_async`
+        // rehydrates this lifetime to be at most as long as the Vmctx's `'a`.
+        let f = unsafe {
+            std::mem::transmute::<LocalBoxFuture<'a, ResumeVal>, LocalBoxFuture<'static, ResumeVal>>(
+                f,
+            )
+        };
+        // Wrap the computation in `YieldedFuture` so that
+        // `Instance::run_async` can catch and run it.  We will get the
+        // `ResumeVal` we applied to `f` above.
+        let ResumeVal(v) = self.yield_val_expecting_val(YieldedFuture(f));
+        // We may now downcast and unbox the returned Box<dyn Any> into an `R`
+        // again.
+        *v.downcast().expect("run_async broke invariant")
+    }
+}
+
+impl Instance {
+    /// Run a WebAssembly function with arguments in the guest context at the
+    /// given entrypoint. Enable `Vmctx::run_await` to trampoline async
+    /// computations back to this context so that we may `.await` on them.
+    ///
+    /// Aside from asynchrony, this function behaves identically to
+    /// `Instance::run`.
+    pub async fn run_async<'a>(
+        &'a mut self,
+        entrypoint: &str,
+        args: &[Val],
+    ) -> Result<RunResult, Error> {
+        if self.is_yielded() {
+            return Err(Error::Unsupported(
+                "cannot run_async a yielded instance".to_owned(),
+            ));
+        }
+
+        // Store the ResumeVal here when we get it.
+        let mut resume_val: Option<ResumeVal> = None;
+        let ret = loop {
+            // Run the WebAssembly program
+            let run = if self.is_yielded() {
+                // A previous iteration of the loop stored the ResumeVal in
+                // `resume_val`, send it back to the guest ctx and continue
+                // running:
+                self._resume_with_val(
+                    resume_val
+                        .take()
+                        .expect("is_yielded implies resume_value is some"),
+                    true,
+                )
```

Comment:
We need to provide a means to run this in [`tokio::task::spawn_blocking`](https://docs.rs/tokio/0.2.21/tokio/task/fn.spawn_blocking.html), but parameterizing this function by that type requires a rank 2 type! Argh.

Reply:
Addressed by fac6457